test-apps/sediment/src/main/java/com/twosigma/waiter/sediment/Servlet.java

```diff
@@ -0,0 +1,222 @@
+        response.getWriter().write(responseContent);
+    }
+
+    void handleTrailersRequest(final Request request, final Response response) throws IOException {
+        final Map<String, Object> infoMap = populateRequestInfo(request);
+        final String responseBody = new Gson().toJson(infoMap);
+        final byte[] outputBytes = responseBody.getBytes();
+
+        final HttpFields trailer = retrieveTrailers(request);
+        if (trailer != null) {
+            response.setTrailers(() -> trailer);
+        }
+
+        final String responseStatusString = request.getHeader("x-sediment-response-status");
+        if (responseStatusString == null) {
+            response.setStatus(200);
+        } else {
+            response.setStatus(Integer.parseInt(responseStatusString));
+        }
+
+        final ByteArrayInputStream respInputStream = new ByteArrayInputStream(outputBytes);
+        final ServletOutputStream respOutputStream = response.getOutputStream();
+        while (true) {
+            final byte[] tempBytes = new byte[128];
+            final int bytesRead = respInputStream.read(tempBytes);
+            if (bytesRead == -1) {
+                final String sleepMsString = request.getHeader("x-sediment-sleep-before-response-trailer-ms");
+                if (sleepMsString != null) {
+                    try {
+                        LOGGER.info("sleeping " + sleepMsString + " ms before returning response trailers");
+                        Thread.sleep(Long.parseLong(sleepMsString));
+                    } catch (InterruptedException ex) {
+                        LOGGER.log(Level.SEVERE, "error while sleeping before sending trailers", ex);
+                    }
+                }
+                break;
+            }
+            respOutputStream.write(tempBytes, 0, bytesRead);
+            respOutputStream.flush();
+        }
+    }
+
+    private HttpFields retrieveTrailers(Request request) {
+        final HttpFields trailerFields = new HttpFields();
+        final String trailerPrefix = "x-sediment-response-trailer-";
+        for (String name : Collections.list(request.getHeaderNames())) {
+            if (name.toLowerCase().startsWith(trailerPrefix)) {
+                final String trailerName = name.substring(trailerPrefix.length());
+                trailerFields.add(trailerName, request.getHeader(name));
+            }
+        }
+        if (trailerFields.getFieldNamesCollection().isEmpty()) {
+            return null;
+        } else {
+            return trailerFields;
+        }
+    }
+
+    Map<String, Object> convertFieldsToMap(final Collection<String> names, final Function<String, Object> retrieveValue) {
+        final Map<String, Object> result = new HashMap<>();
+        for (String name : names) {
+            result.put(name, retrieveValue.apply(name));
+        }
+        return result;
+    }
+
+    Map<String, Object> populateRequestInfo(final Request request) throws IOException {
```

Comment:
The goal of this function seems to limited to creating the response JSON.  It seems better to directly use a [JsonWriter](https://static.javadoc.io/com.google.code.gson/gson/2.6.2/com/google/gson/stream/JsonWriter.html) created on the response's output stream, avoiding the `ByteArrayInputStream`, etc.